### ISSUE TITLE
Middeware specification

### ISSUE THREAD

**Issue by tegefaulkes** [MatrixAI/Polykey#502]
### Specification

Still fleshing out the concept.

The middleware is shared between the Client and Server side implementation. So I'm making a single issue for just the middlewear.

**Comment by tegefaulkes** [comment]
The middlewear can be implemented as a generator. it will operate over the data of the messages. It should be able to output metadata for the output stream. There are two options here. Use a transformation stream to feed the input messages to the interceptor generator. The readable stream can alternatively be `tee()`'d into a writable stream that feeds the generators. The output of the generators can be combined together for output. 

One point of fuzziness is that we'd have to add metadata to the outgoing messages, but incoming and outgoing messages are not 1:1. I don't think it's a problem to have no metadata for some outgoing messages but how do we handle multiple incoming metadata per 1 outgoing metadata? we can't exactly combine the metadata the usual way. what would loose some information if they share object keys.  do we provide an array of the metadata's waiting for output? Doesn't the middlewear run for every incoming message or just ones with metadata?

So the flow is something like this. incoming message -> middlewear -> middlewear response buffer -> output message.

A transformation stream can be used to take outgoing metadata from a buffer and add it to the output messages.

I think this design is mostly functional. But it still seems possible to end a stream with some metadata waiting to be sent. What do we do in this case? Do we construct a message without and data content and flush it out? Do we by convention force a null message to signal end of the handling? That would be the same as automatically making a null message just for the metadata. Do we just ignore the problem and give no guarantee that the last metadata will be sent?

**Comment by CMCDragonkai** [comment]
For our uses of authentication metadata, we would generally be sending metadata on the initial request and the initial response.

I don't think there's any other situation where must have metadata on the trailing response or trailing request.

**Comment by tegefaulkes** [comment]
We want to replicate the usual middlewear style of `rpc.use((req, res, next) => void)`. In this case we want to use form a transformation stream chain that handles the middlewear. The middlewear will act on the whole JSONRPC messages.

For simplicity sake I think we'll need to register the forward and reverse path separately. If we wanted a forward middleware to modify the reverse messages then it will be a little complex. It should be possible to combine two transformation streams to short-circuit some of the data path.

I need to do a bunch of prototyping for this.

**Comment by tegefaulkes** [comment]
I've prototyped out the middle-ware using transformation streams and I have something workable here but it's pretty basic.

Essentially I've added the ability to register forward and reverse middlewear. When you register a middlewear you provide a factory function that will return the middleware with it's initialized state. 

```ts
type Middleware<T> = (input: ReadableStream<T>) => ReadableStream<T>;
type MiddlewareFactory<T> = () => Middleware<T>
```

The Middleware itself is a function that takes a readable stream and returns the modified readable stream. It is basically acting as a transformation stream. This gives us the freedom to read or modify the complete message.

The simplest way to use this is to take the input and use a `TransformationStream` with `.pipeThrough()`. But you have the freedom do do anything so long as you return a `ReadableStream`.

```ts
(input) => input.pipeThrough(new TransformationStream())
```

Currently I don't have a way of short circuiting the output. I'll need to add that if we want a simple way of bypassing the handlers. We'll need this for authentication where we'd throw an error to the handler and immediately return and authentication error message.

**Comment by CMCDragonkai** [comment]
If you want to short-circuit the output, it would mean doing it before propagating the first mesesage to the output stream (the next middleware). That would enable you ensure that nothing could have already been sent on the output stream at the middleware layer you're operating in.

**Comment by tegefaulkes** [comment]
I've got short-circuiting working. The middleware is provided a function called `short` (for now) that you can call and return a response message. This will push the message into a queue. An output transformer will consume the queue first before outputing any of the incoming messages. This means the short circuit messages are sent through first. If you want to stop the handler at any point you throw an error in your middle ware transform. Note that this error is advisory, the handler can catch and ignore it. I can add a way to kill the output stream if it's needed.

The design so far is workable but it needs a review and maybe some touch ups. I'm going to implement the middleware on the RPCClient now.

**Comment by CMCDragonkai** [comment]
I think you should avoid using exceptions for expected control flow like that. Surely there must be a better way to short-circuit.

**Comment by tegefaulkes** [comment]
The error is only to force the handler to end. It might be cleaner to terminate the stream? In that case the writer side will throw but the reader side will just close early. We may have to terminate the forward and reverse path if we want to end the handler here. The problem is, I don't see a clean way of doing this without the handler possibly ending up with an error. 

For example, if authentication fails and we terminate both streams, currently the handler logic will throw an error due to the stream ending before the first message. If that didn't happen then it would throw trying to write to the output. If it's going to  throw anyway then I'd prefer to see a clear error of 'authentication failed' rather than 'stream ended early' or 'failed to write, writer closed' (I'm paraphrasing the errors here).

So basically, if the middleware is going to terminate the handler early I'd prefer a clear error message, middleware ending the streams would typically be an 'exceptional' exception such as failures in authentication, encryption, compression, etc, etc.

**Comment by tegefaulkes** [comment]
The middleware implementation is pretty much done now. All that is left is a general review and any fixes.

I still need to write out the spec based on the current implementation.
